pyathena/common.py

```diff
@@ -142,24 +142,64 @@ def _build_start_query_execution_request(self, query, work_group=None, s3_stagin
             })
         return request
 
-    def _execute(self, operation, parameters=None, work_group=None, s3_staging_dir=None):
+    def _find_previous_query_id(self, query, work_group, cache_size):
+        try:
+            next_token = None
+            while cache_size > 0:
+                n = min(cache_size, 50)  # 50 is max allowed by AWS API
+                cache_size -= n
+                request = {'MaxResults': n}
+                if work_group is not None:
+                    request['WorkGroup'] = work_group
+                if next_token is not None:
+                    request['NextToken'] = next_token
+                response = retry_api_call(self.connection._client.list_query_executions,
+                                          config=self._retry_config,
+                                          logger=_logger,
+                                          **request)
+                query_ids = response.get('QueryExecutionIds', None)
+                if not query_ids:
+                    break  # no queries left to check
+                next_token = response.get('NextToken', None)
+                query_executions = retry_api_call(
+                    self.connection._client.batch_get_query_execution,
+                    config=self._retry_config,
+                    logger=_logger,
+                    QueryExecutionIds=query_ids
+                ).get('QueryExecutions', [])
+                for execution in query_executions:
+                    queries_match = execution['Query'] == query
+                    success = execution['Status']['State'] == AthenaQueryExecution.STATE_SUCCEEDED
+                    is_dml = execution['StatementType'] == 'DML'
```

Comment:
I thought there was no reason to assign it to a variable.

Reply:
I find it more readable than a multi-line if statement, but 🤷‍♂